### PR TITLE
Use scikit build instead of setup.py

### PR DESCRIPTION
**Pull request by josephmckinsey**
- Default behavior is to download helics from src and build.
- If DOWNLOAD_BINARIES or CIBUILDWHEEL are defined, then it downloads the binaries from the repository.
- There is not a way to download locally, even with an sdist. I couldn't quite figure out how to get scikit-build to use a local version of scikit build core along with its other dependencies.
- Setting sdist always downloads helics-src.
- Requires libzqm and boost to build from source. These are necessary for pytest, so I figured we should include them.
- Switches to dynamic versioning. The `setuptools-scm` does not seem to be required, but @nightlark mentioned using the dynamic-version tag. What this means is that `helics/_version.py` has been deleted from git, and instead it is automatically generated from the latest git tag. This makes `.github/workflows/update-helics.yml` redundant.

## Remaining items

- [x] Fix .github workflows to use cibuildwheel (and test resulting wheels)
- [ ] Get offline builds working

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+if(SKBUILD_STATE STREQUAL "wheel")
+    message(WARNING "Building HELICS from source (some features including ZeroMQ Core, IPC Core, and web server won't be available)")
+endif()
+
+set(CMAKE_CXX_STANDARD 17)
+set(HELICS_DISABLE_GIT_OPERATIONS ON)
+set(HELICS_ZMQ_SUBPROJECT ON)
+set(HELICS_ZMQ_FORCE_SUBPROJECT ON)
+set(HELICS_ENABLE_ZMQ_CORE ON) # TODO: vendor libzmq similar to other 3rd party libraries
+set(HELICS_DISABLE_BOOST OFF) # TODO: headers are just needed to compile, maybe tell users to have it installed as prereq for building from source?
```
Comment (nightlark) [inline CMakeLists.txt:81]: I think we should users to have it installed as a pre-requisite for buidling from source. I think we need to update the pyproject.toml file to exclude the copy of boost that was downloaded (or have cibuildwheel on Linux stick it in a folder outside of the repository clone). A copy of Boost has made its way into the current source dist, making it take up about 250MB compressed -- or 1+ GB uncompressed.

File: pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+Homepage = "https://github.com/GMLC-TDC/pyhelics"
+Discussions = "https://github.com/GMLC-TDC/HELICS/discussions"
+Documentation = "https://python.helics.org/"
+"Issue Tracker" = "https://github.com/GMLC-TDC/pyhelics/issues"
+"Source Code" = "https://github.com/GMLC-TDC/pyhelics"
+
+[tool.cibuildwheel]
+test-command = ["helics --version"]
+linux.manylinux-x86_64-image = "manylinux_2_28"
+linux.manylinux-aarch64-image = "manylinux_2_28"
```
[the remark this answers is not in the document]
Comment (josephmckinsey) [reply pyproject.toml:69]: https://github.com/pypa/manylinux?tab=readme-ov-file#manylinux_2_24-debian-9-based---eol

That docker container isn't being maintained anymore.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -1,6 +1,12 @@
 name: CI
 
-on: [push, pull_request]
+on:
+  push:
+    branches:
+    - main
+  pull_request:
+    branches:
+    - main
```
[the remark this answers is not in the document]
Comment (josephmckinsey) [reply .github/workflows/ci.yml:9]: Done.

File: pyproject.toml (hunk 2 above)
[the remark this answers is not in the document]
Comment (nightlark) [reply pyproject.toml:69]: Are they still maintaining `manylinux2014`, despite the OS used being EOL?

[the remark this answers is not in the document]
Comment (nightlark) [reply pyproject.toml:69]: It looks like they're supporting `manylinux2014` for another year: https://github.com/pypa/manylinux/discussions/1648

File: CMakeLists.txt (hunk 1 above)
Comment (josephmckinsey) [reply CMakeLists.txt:81]: I've moved it to /tmp/boost. I'm still not sure why it got included in the sdist.

File: CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,101 @@
+cmake_minimum_required(VERSION 3.15...3.26)
```
[the remark this answers is not in the document]
Comment (josephmckinsey) [reply CMakeLists.txt:1]: > [cmake_minimum_required(VERSION)](https://cmake.org/cmake/help/v3.20/command/cmake_minimum_required.html#command:cmake_minimum_required) command implicitly calls cmake_policy(VERSION)

File: .github/workflows/pythonpackage.yml [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+    inputs:
+      use_qemu:
+        description: 'Use qemu to build linux aarch64, ppc64le & s390x'
+        required: false
+        default: 'false'
+  workflow_run:
+    workflows: ["HELICS Version Update"]
+    branches: [main]
+    types:
+      - completed
```
[the remark this answers is not in the document]
Comment (josephmckinsey) [reply .github/workflows/pythonpackage.yml:14]: How does a new branch gets tagged right now? This doesn't happen in update helics version, right?

If I understand you correctly, you mean
1. Update HELICS version gets run.
2. Run tests and build python package
3. Make a tag if tests pass
4. Upload repository by calling pythonpackage agains.

File: pyproject.toml (hunk 2 above)
[the remark this answers is not in the document]
Comment (josephmckinsey) [reply pyproject.toml]: > Error: Command ['sh', '-c', "yum install -y wget && mkdir /boost && pushd /boost && wget --no-check-certificate 'https://sourceforge.net/projects/boost/files/boost/1.76.0/boost_1_76_0.tar.bz2' &&  tar xf boost_1_76_0.tar.bz2 && popd\n"] failed with code 1.

**Review by nightlark — APPROVED** [review]

**Comment by josephmckinsey** [comment]
@nightlark build-helics-sdist seems to no longer find boost anymore. Was there a change in how we find boost in HELICS?

**Comment by nightlark** [comment]
The minimum boost version required was bumped up. It might try to use the presence of the CMake Config package for Boost more frequently?

**Comment by nightlark** [comment]
These are the relevant lines for finding Boost on Linux systems: https://github.com/GMLC-TDC/HELICS/blob/main/config/cmake/addBoost.cmake#L110-L182

There was a change to make the handling of the Boost_ROOT environment variables more consistent with the documentation of the CMake documentation for finding Boost.

**Comment by nightlark** [comment]
Okay, there were a lot of changes in that file, made in this commit to make it use more of the CMake provided code for finding Boost libraries: https://github.com/GMLC-TDC/HELICS/commit/194ec8559#diff-a6960b231bb7ad1c7a695be7cf7e90db6ebbf9b30ee3fcc35ae2bba1ce4b7b5b

I think the key thing is make sure your Boost Root variable is set to a folder has one of these exact paths as subfolders: `boost/version.hpp boost/config.hpp`

**Comment by josephmckinsey** [comment]
> Okay, there were a lot of changes in that file, made in this commit to make it use more of the CMake provided code for finding Boost libraries: [GMLC-TDC/HELICS@194ec8559#diff-a6960b231bb7ad1c7a695be7cf7e90db6ebbf9b30ee3fcc35ae2bba1ce4b7b5b](https://github.com/GMLC-TDC/HELICS/commit/194ec8559#diff-a6960b231bb7ad1c7a695be7cf7e90db6ebbf9b30ee3fcc35ae2bba1ce4b7b5b)
> 
> I think the key thing is make sure your Boost Root variable is set to a folder has one of these exact paths as subfolders: `boost/version.hpp boost/config.hpp`

Thanks. I figured that out right before you sent it. The aarch64 build takes its sweet time for some reason, but it looks to be working.

**Pull request merged**